File: src/main/java/net/gegy1000/terrarium/server/world/layer/ReplaceRandomLayer.java
package net.gegy1000.terrarium.server.world.layer;

import net.minecraft.world.gen.layer.GenLayer;
import net.minecraft.world.gen.layer.IntCache;

public class ReplaceRandomLayer extends GenLayer {
    private final int replace;
    private final int replacement;
    private final int chance;

    public ReplaceRandomLayer(int replace, int replacement, int chance, long seed, GenLayer parent) {
        super(seed);
        this.replace = replace;
        this.replacement = replacement;
        this.chance = chance;
        this.parent = parent;
    }

    @Override
    public int[] getInts(int areaX, int areaY, int areaWidth, int areaHeight) {
        int[] parent = this.parent.getInts(areaX, areaY, areaWidth, areaHeight);
        int[] result = IntCache.getIntCache(areaWidth * areaHeight);
        for (int z = 0; z < areaHeight; z++) {
            for (int x = 0; x < areaWidth; x++) {
                int index = x + z * areaWidth;
                int sample = parent[index];
                result[index] = this.shouldReplace(areaX + x, areaY + z, sample) ? this.replacement : sample;
            }
        }
        return result;
    }

    private boolean shouldReplace(long x, long z, int sample) {
        if (sample == this.replace) {
            this.initChunkSeed(x, z);
            return this.nextInt(this.chance) == 0;
        }
        return false;
    }
}

File: src/main/kotlin/net/gegy1000/terrarium/server/map/glob/generator/layer/ReplaceRandomLayer.kt
package net.gegy1000.terrarium.server.map.glob.generator.layer

import net.minecraft.world.gen.layer.GenLayer
import net.minecraft.world.gen.layer.IntCache

class ReplaceRandomLayer(val replace: Int, val replacement: Int, val chance: Int, seed: Long, parent: GenLayer) : GenLayer(seed) {
    init {
        this.parent = parent
    }

    override fun getInts(areaX: Int, areaY: Int, areaWidth: Int, areaHeight: Int): IntArray {
        val parent = this.parent.getInts(areaX, areaY, areaWidth, areaHeight)
        val result = IntCache.getIntCache(areaWidth * areaHeight)
        for (z in 0..areaHeight - 1) {
            for (x in 0..areaWidth - 1) {
                this.initChunkSeed((areaX + x).toLong(), (areaY + z).toLong())
                val index = x + z * areaWidth
                val sample = parent[index]
                if (sample == this.replace && this.nextInt(this.chance) == 0) {
                    result[index] = this.replacement
                } else {
                    result[index] = sample
                }
            }
        }
        return result
    }
}
